cmd/restic/cmd_prune.go

```diff
@@ -56,207 +101,348 @@ func runPrune(gopts GlobalOptions) error {
+			stats.blobs.unused++
 		}
+		// update indexPack
+		indexPack[blob.PackID] = ip
 	}
 
-	removeBytes := duplicateBytes
-
-	// find packs that are unneeded
+	Verbosef("collect packs for deletion and repacking...\n")
+	removePacksFirst := restic.NewIDSet()
 	removePacks := restic.NewIDSet()
+	repackPacks := restic.NewIDSet()
 
-	Verbosef("will remove %d invalid files\n", len(invalidFiles))
-	for _, id := range invalidFiles {
-		removePacks.Insert(id)
+	var repackCandidates []packInfoWithID
+
+	repack := func(id restic.ID, p packInfo) {
+		repackPacks.Insert(id)
+		stats.blobs.repack += p.unusedBlobs + p.duplicateBlobs + p.usedBlobs
+		stats.size.repack += p.unusedSize + p.usedSize
+		stats.blobs.repackrm += p.unusedBlobs
+		stats.size.repackrm += p.unusedSize
 	}
 
-	for packID, p := range idx.Packs {
+	// loop over all packs and decide what to do
+	bar := newProgressMax(!gopts.Quiet, uint64(len(indexPack)), "packs processed")
+	bar.Start()
+	err := repo.List(ctx, restic.PackFile, func(id restic.ID, packSize int64) error {
+		p, ok := indexPack[id]
+		if !ok {
+			// Pack was not referenced in index and is not used  => immediately remove!
+			Verbosef("pack %s is not referenced in any index and not used -> will be removed.\n", id.Str())
+			removePacksFirst.Insert(id)
+			stats.size.unref += uint64(packSize)
+			return nil
+		}
 
-		hasActiveBlob := false
-		for _, blob := range p.Entries {
-			h := restic.BlobHandle{ID: blob.ID, Type: blob.Type}
-			if usedBlobs.Has(h) {
-				hasActiveBlob = true
-				continue
-			}
+		if p.unusedSize+p.usedSize != uint64(packSize) {
+			Verbosef("pack %s: calculated size %d does not match real size %d difference: %d => treated as mixed pack\n",
+				id.Str(), p.unusedSize+p.usedSize, packSize, packSize-int64(p.unusedSize+p.usedSize))
+			p.tpe = restic.InvalidBlob
+		}
 
-			removeBytes += uint64(blob.Length)
+		// statistics
+		switch {
+		case p.tpe == restic.InvalidBlob:
+			stats.packs.mixed++
+		case p.tpe == restic.TreeBlob:
+			stats.packs.tree++
+		default:
+			stats.packs.data++
+		}
+		switch {
+		case p.usedBlobs == 0 && p.duplicateBlobs == 0:
+			stats.packs.unused++
+		case packSize < repository.MinPackSize:
+			stats.packs.small++
+		case p.unusedBlobs == 0:
+			stats.packs.used++
+		default:
+			stats.packs.partlyUsed++
 		}
 
-		if hasActiveBlob {
-			continue
+		// decide what to do
+		switch {
+		case p.usedBlobs == 0 && p.duplicateBlobs == 0:
+			// All blobs in pack are no longer used => remove pack!
+			removePacks.Insert(id)
+			stats.blobs.remove += p.unusedBlobs
+			stats.size.remove += p.unusedSize
+
+		case opts.RepackTreesOnly && p.tpe != restic.TreeBlob,
+			// if this is a data pack and --repack-trees-only is set => keep pack!
+			uint(len(repackPacks)) >= opts.MaxRepackCount:
+			// also keep if number of packs to repack is reached
+			stats.packs.keep++
+
+		case opts.RepackMixed && p.tpe == restic.InvalidBlob,
+			opts.RepackSmall && packSize < repository.MinPackSize,
```

Comment:
Actually I observed packs which are almost 12MB. This can be explained as blobs can have a size up to 8MB. So if the pack is almost full (shortly below 4MB) and a 8MB blob is added, you get this packsize.
Altogether, I would not say that `MinPackSize` is the normal pack size and I would claim all packs smaller that that size "small" packs.

About the single data or tree pack file: Of course, we could count tree and data pack files separately. But this would make the code even more complex. If you wish, I can add this.

Reply:
TreePacks can actually be far larger than 12MB. The average packsize is probably still in the 4-5MB range. I had in mind that the closer a small pack is to `MinPackSize` the less we gain in terms of reducing the number pack files. Its basically a tradeoff between fewer files and less work with repacking.

Making the code even more complex doesn't sound good. The small useless repack is probably not worth the trouble.

Reply:
As this option is now commented out, I'll mark this point as resolved.